sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -304,7 +329,13 @@ case class DataSource(
       case (dataSource: SchemaRelationProvider, Some(schema)) =>
         dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions, schema)
       case (dataSource: RelationProvider, None) =>
-        dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
+        val baseRelation =
+          dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
+        SchemaUtils.checkColumnNameDuplication(
+          baseRelation.schema.map(_.name),
+          "in the relation schema",
+          equality)
+        baseRelation
```

Comment:
an idea to unify the validation for both user-specified schema and inferred schema: we can put the validation at the end of `resolveRelation`:
```
relation match {
  case h: HadoopFsRelation =>
    // validate data schema
    // validate partition schema
    // validate bucket
  case _ => // validate data schema
}
relation
```